### ISSUE TITLE
[BUG] docker compose up failing to connect to postgres container on Apple Silicon Macs

### ISSUE THREAD

**Issue by taimurshaikh** [docker/compose#10673]
### Description

# Current Behavior
When running `docker compose up` to set up a dev environment, my custom containers depend on a `postgres` container. On most attempts of running `docker compose up`, the custom containers fail to make a connection with `postgres`, however there are occasions when it works. This issue is occuring for two out of three members of our team with Apple Silicon machines (M1 and M2), but it does not occur for the member of our team with an Intel Mac. We have already tried switching to the `arm64v8/postgres` version of the postgres image but that has not fixed the issue.
``` yaml
version: '3'
services:
  postgres:
    image: postgres
    environment:
      - POSTGRES_USER=REDACTED
      - POSTGRES_PASSWORD=REDACTED
      - POSTGRES_DB=REDACTED
    # volumes:
    #   - ./learn_prisma_default:/var/lib/docker/volumes/learn_prisma_postgres/_data
  redis:
    image: redis
    env_file:
      - .env

# My two custom containers
  venture-copilot:
    image: venture-copilot
    ports:
      - "8080:8080"
    env_file:
      - .env
    build:
      context: .
      dockerfile: Dockerfile
    depends_on:
      - postgres
      - redis

  dashboard:
    image: dashboard
    ports:
      - "80:80"
    env_file:
      - .env
    build:
      context: dashboard
      dockerfile: Dockerfile
    depends_on:
      - postgres
      - redis
  ```
Here are the logs from a successful and failed example: https://gist.github.com/taimurshaikh/c497167b00c4bc06c05fded18ca9cd0e

### Compose Version

```Text
Docker Compose version v2.12.2
```

### Docker Environment

```Text
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.9.1)
  compose: Docker Compose (Docker Inc., v2.12.2)
  dev: Docker Dev Environments (Docker Inc., v0.0.3)
  extension: Manages Docker extensions (Docker Inc., v0.2.13)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scan: Docker Scan (Docker Inc., v0.21.0)
WARNING: Plugin "/Users/taimurshaikh/.docker/cli-plugins/docker-init" is not valid: failed to fetch metadata: fork/exec /Users/taimurshaikh/.docker/cli-plugins/docker-init: no such file or directory
WARNING: Plugin "/Users/taimurshaikh/.docker/cli-plugins/docker-scout" is not valid: failed to fetch metadata: fork/exec /Users/taimurshaikh/.docker/cli-plugins/docker-scout: no such file or directory

Server:
 Containers: 4
  Running: 0
  Paused: 0
  Stopped: 4
 Images: 4
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2 io.containerd.runtime.v1.linux
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1c90a442489720eec95342e1789ee8a5e1b9536f
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.49-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 4
 Total Memory: 7.667GiB
 Name: docker-desktop
 ID: 45IS:TCZX:YB3F:E5KM:ZGAU:SVYO:CEAW:YNSQ:IIPF:QYK4:5P2V:QIIT
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5000
  127.0.0.0/8
 Live Restore Enabled: false
```

### Anything else?
List of things we have tried so far:
* Adding a healthcheck to the postgres container, experimenting with different timeout lengths and number of retries
* Upgrading Docker Desktop to latest version

**Comment by glours** [comment]
Hello @taimurshaikh 

Can you try to add an `healthcheck` attribute to your `postgres` service, and a `condition` attribute to your `depends_on` like this: 
```
version: '3'
services:
  postgres:
    image: postgres
    environment:
      - POSTGRES_USER=REDACTED
      - POSTGRES_PASSWORD=REDACTED
      - POSTGRES_DB=REDACTED
    healthcheck:
      test: [ "CMD", "pg_isready" ]
      interval: 10s
      timeout: 5s
      retries: 5
    # volumes:
    #   - ./learn_prisma_default:/var/lib/docker/volumes/learn_prisma_postgres/_data
  redis:
    image: redis
    env_file:
      - .env

# My two custom containers
  venture-copilot:
    image: venture-copilot
    ports:
      - "8080:8080"
    env_file:
      - .env
    build:
      context: .
      dockerfile: Dockerfile
    depends_on:
      postgres:
        condition: service_healthy
      redis:

  dashboard:
    image: dashboard
    ports:
      - "80:80"
    env_file:
      - .env
    build:
      context: dashboard
      dockerfile: Dockerfile
    depends_on:
      postgres:
        condition: service_healthy
      redis:
```

**Comment by taimurshaikh** [comment]
Hi @glours thanks for your quick response. Upon implementing your suggestions, we still face the same issue, and also this new error message appears in the logs continually:
```
venture-copilot-app-postgres-1         | 2023-06-08 15:21:02.790 UTC [95] FATAL:  role "root" does not exist
```

**Comment by glours** [comment]
Ho my bad, I didn't notice your custom user
change the check line with ```test: [ "CMD", "pg_isready -d $${POSTGRES_DB} -U $${POSTGRES_USER}" ]```

**Comment by taimurshaikh** [comment]
It appears the healtcheck fails:
```container for service "postgres" is unhealthy```
I tried experimenting with the number of retries and the timeout but it seems something is wrong with how the app is responding with `postgres`

**Comment by milas** [comment]
It looks like you're running Docker Desktop 4.14.0, which is from November 2022.

Can you upgrade to the latest version (`Docker Desktop > Settings > Software Updates > Check for updates`) and see if that helps?

**Comment by taimurshaikh** [comment]
The issue persisted when I had the most recent version of Docker Desktop installed - our team tried a universal downgrade to the version that was used when our codebase was originally set up but unfortunately that was no good either. I will upgrade again though on my machine and let you know if there are any changes.

**Comment by glours** [comment]
@taimurshaikh can you share the log of your postgres service? It seems you have either a postgres configuration issue or, as you didn't tag your version of postgres, you're now using a version of postgres incompatible with your previous config.
I did a `docker image inspect postgres` to check what is the current version of the latest flag and this is a `PG_VERSION=15.3-1.pgdg110+1`. Is it the version you expect to use?

**Comment by taimurshaikh** [comment]
Yes this is the one that works for out other team member with an Intel Mac which leads me to believe its a processor-related issue. I have seen examples of docker related issues when it comes to Apple Silicon e.g. on this page for Argilla https://docs.argilla.io/en/latest/getting_started/quickstart_installation.html

**Comment by glours** [comment]
Can you try to use this option in Docker Desktop (Settings > Features in development) to see if this could fix your issue
<img width="710" alt="Screenshot 2023-06-09 at 15 34 20" src="https://github.com/docker/compose/assets/705411/3c9aed86-c798-41bf-a4fc-18b18f0b1220">

**Comment by glours** [comment]
And regarding the platform issue mentioned in the argilla doc, Compose will pull by default the image for your specific platform, so if a `arm/64` version is available it will be used

**Comment by taimurshaikh** [comment]
The option to use Rosetta is not showing up in my settings page even after enabling the virtualization framework and restarting the app. I will restart my machine and see if that does anything

**Comment by tushar5526** [comment]
Yes, I can confirm this is happening with me as well. Randomly my connections to postgres containers are failing. I tried running the same setup on github codespaces and it worked fine on multiple tries. I am on Macbook Pro Apple Silicon M2 chip.

**Comment by kelvin-lima** [comment]
Additionally to the last message... 
After deleting the problematic container, in this case `Redis`, and running compose again I got:
```Error response from daemon: network d4cfffd6bcba7c9a7eb1060e31c90b754207385a526c62cba28cb4b071574595 not found```

After that, running `docker network prune`, which deleted the `redis_api` and running the compose again, the container spin up and both services are connected now. 

But If I remove all containers and spin up again, the problem happens.

This started happening after updating the Docker to:

![Screenshot 2023-07-20 at 13 36 18](https://github.com/docker/compose/assets/3374682/b4d407f8-a3ca-4ce1-921e-06e4ef537051)

**Comment by RockingThor** [comment]
I'm getting the same error as previously mentioned. Docker is not creating the `postgres` container. It is just creating the `app` build. As my `app` depends on the `postgres` the build is getting failed everytime.

Here is my docker-compose file:
`services:
  app:
    build:
      context: ./
      dockerfile: Dockerfile
    volumes:
      - .:/app
    ports:
      - "3000:3000"
    depends_on:
      - postgres
    networks:
      - app_network
  postgres:
    image: "postgres:latest"
    container_name: postgres
    restart: always
    ports:
      - "5432:5432"
    volumes:
      - "pgdata:/var/lib/postgresql/data/"
    networks:
      - app_network
    environment:
      POSTGRES_USER: rohitnandi
      POSTGRES_PASSWORD: postgres
networks:
  app_network:
    driver: bridge
volumes:
  pgdata:`

Here is the error I'm getting:

`docker compose up  
[+] Building 5.9s (14/14) FINISHED                                                                                                                                                                                                                       docker:desktop-linux
 => [app internal] load build definition from Dockerfile                                                                                                                                                                                                                 0.0s
 => => transferring dockerfile: 817B                                                                                                                                                                                                                                     0.0s
 => [app internal] load .dockerignore                                                                                                                                                                                                                                    0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                          0.0s
 => [app internal] load metadata for docker.io/library/node:20                                                                                                                                                                                                           2.8s
 => [app auth] library/node:pull token for registry-1.docker.io                                                                                                                                                                                                          0.0s
 => [app 1/9] FROM docker.io/library/node:20@sha256:5f21943fe97b24ae1740da6d7b9c56ac43fe3495acb47c1b232b0a352b02a25c                                                                                                                                                     0.0s
 => [app internal] load build context                                                                                                                                                                                                                                    2.0s
 => => transferring context: 4.44MB                                                                                                                                                                                                                                      1.8s
 => CACHED [app 2/9] RUN apt-get update -y && apt-get install -y openssl                                                                                                                                                                                                 0.0s
 => CACHED [app 3/9] WORKDIR /app                                                                                                                                                                                                                                        0.0s
 => CACHED [app 4/9] COPY package*.json ./                                                                                                                                                                                                                               0.0s
 => CACHED [app 5/9] RUN npm install                                                                                                                                                                                                                                     0.0s
 => CACHED [app 6/9] COPY .. .                                                                                                                                                                                                                                           0.0s
 => CACHED [app 7/9] RUN npx prisma generate                                                                                                                                                                                                                             0.0s
 => CACHED [app 8/9] RUN sleep 1                                                                                                                                                                                                                                         0.0s
 => ERROR [app 9/9] RUN npx prisma db push                                                                                                                                                                                                                               1.0s
------
 > [app 9/9] RUN npx prisma db push:
0.873 Environment variables loaded from .env
0.873 Prisma schema loaded from prisma/schema.prisma
0.876 Datasource "db": PostgreSQL database "postgres", schema "public" at "postgres:5432"
0.921 
0.921 Error: P1001: Can't reach database server at `postgres`:`5432`
0.921 
0.921 Please make sure your database server is running at `postgres`:`5432`.
------
failed to solve: process "/bin/sh -c npx prisma db push" did not complete successfully: exit code: 1`

**Comment by glours** [comment]
Hey everyone 👋 
@kelvin-lima @taimurshaikh @tushar5526 
Can you test again with a recent version of Compose? We fixed a couple a network problems since this issue was opened, especially this [one] (https://github.com/docker/compose/pull/10778) in Compose `v2.20.0`

@RockingThor you should add a `healthcheck` configuration to your postgres service, [see here for an example](https://github.com/docker/compose/issues/10673#issuecomment-1582766075), and the following `depends_on` config to your `app` service
```
depends_on:
      postgres:
        condition: service_healthy
```
 This way you'll be sure your `app` service to wait until your pg service is ready to accept connection

**Comment by RockingThor** [comment]
> Hey everyone 👋 @kelvin-lima @taimurshaikh @tushar5526 Can you test again with a recent version of Compose? We fixed a couple a network problems since this issue was opened, especially this [one] (#10778) in Compose `v2.20.0`
> 
> @RockingThor you should add a `healthcheck` configuration to your postgres service, [see here for an example](https://github.com/docker/compose/issues/10673#issuecomment-1582766075), and the following `depends_on` config to your `app` service
> 
> ```
> depends_on:
>       postgres:
>         condition: service_healthy
> ```
> 
> This way you'll be sure your `app` service to wait until your pg service is ready to accept connection

Thanks for your attention to the issue. Tried exactly what you instructed still getting the same error as my app build is unable to connect to the postgres container.
Here is my updated docker-compose file:

`services:
  app:
    build:
      context: ./
      dockerfile: Dockerfile
    volumes:
      - .:/app
    ports:
      - "3000:3000"
    depends_on:
      postgres:
        condition: service_healthy
    networks:
      - app_network
  postgres:
    image: "postgres:latest"
    container_name: postgres
    restart: always
    ports:
      - "5432:5432"
    volumes:
      - "pgdata:/var/lib/postgresql/data/"
    networks:
      - app_network
    environment:
      POSTGRES_USER: rohitnandi
      POSTGRES_PASSWORD: postgres
    healthcheck:
      test: ["CMD", "pg_isready"]
      interval: 10s
      timeout: 5s
      retries: 5
networks:
  app_network:
volumes:
  pgdata:`

This is my database URL:
`POSTGRES_PRISMA_URL=postgresql://rohitnandi:postgres@postgres:5432`

Here is the error message:
` => [app 7/8] RUN npx prisma generate                                                                                                                                                                                             2.0s
 => ERROR [app 8/8] RUN npx prisma db push                                                                                                                                                                                        0.8s 
------                                                                                                                                                                                                                                 
 > [app 8/8] RUN npx prisma db push:
0.701 Environment variables loaded from .env
0.702 Prisma schema loaded from prisma/schema.prisma
0.705 Datasource "db": PostgreSQL database "postgres", schema "public" at "postgres:5432"
0.776 
0.776 Error: P1001: Can't reach database server at `postgres`:`5432`
0.776 
0.776 Please make sure your database server is running at `postgres`:`5432`.
------
failed to solve: process "/bin/sh -c npx prisma db push" did not complete successfully: exit code: 1`

**Comment by tushar5526** [comment]
Hey @glours thanka for the awesome product, this got fixed I suppose a few weeks after I faced this issue, maybe in one of the docker updates. Not encountered the issue anywhere else so far.

**Comment by milas** [comment]
@RockingThor Compose services/dependencies determine the order services are launched with `compose up`. Your _image build_ for `app` is trying to use Postgres, which won't be available - even if the service is running, the build sandbox won't have access to it. That's not something that's supported by Compose.

--

Closing this issue - as mentioned, the original cause of these issues and other sporadic networking problems was resolved in #10778, so the solution is to make sure you're on the latest version of Docker Desktop (or Compose itself if you manually installed it).
